kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java

```diff
@@ -0,0 +1,489 @@
+        ),
+        new FindOneAndUpdateOptions()
+            .upsert(true)
+            .returnDocument(ReturnDocument.AFTER)
+    );
+
+    if (metaDoc == null) {
+      throw new IllegalStateException("Uninitialized metadata for partition " + kafkaPartition);
+    }
+
+    LOG.info("{}[{}] Retrieved initial metadata {}", name, kafkaPartition, metaDoc);
+
+    final var partitionSegments = new PartitionSegments(metaDoc.streamTime, metaDoc.epoch);
+
+    final var activeSegments = partitioner.activeSegments(kafkaPartition, metaDoc.streamTime);
+    if (activeSegments.isEmpty()) {
+      LOG.info("{}[{}] No active segments for initial streamTime {}",
+               name, kafkaPartition, metaDoc.streamTime);
+    } else {
+      for (final SegmentPartition segmentToCreate : activeSegments) {
+        // Most likely if we have active segments upon initialization, the corresponding
+        // collections and windowStart indices already exist, in which case #getCollection and
+        // #createIndex should be quick and won't rebuild either structure if one already exists
+        // TODO: optimize by using listCollectionNames to skip redundant collection/index creation
+        //  since we should almost always have an index for each collection for all active segments
+        partitionSegments.createSegment(segmentToCreate);
+      }
+
+      final long firstSegmentId = activeSegments.get(0).segmentId;
+      LOG.info("{}[{}] Initialized active segments in range {} - {}",
+               name, kafkaPartition, firstSegmentId, firstSegmentId + activeSegments.size());
+    }
+
+    kafkaPartitionToSegments.put(
+        kafkaPartition,
+        partitionSegments
+    );
+
+    return new MongoWindowFlushManager(
+        this,
+        (segment) -> windowsForSegmentPartition(kafkaPartition, segment),
+        metadata,
+        partitioner,
+        kafkaPartition
+    );
+  }
+
+  public String collectionNameForSegment(final SegmentPartition segmentPartition) {
+    return String.format(
+        "%s-%d-%d",
+        name, segmentPartition.tablePartition, segmentPartition.segmentId
+    );
+  }
+
+  public MongoCollection<WindowDoc> windowsForSegmentPartition(
+      final int kafkaPartition,
+      final SegmentPartition segment
+  ) {
+    return kafkaPartitionToSegments.get(kafkaPartition).segmentWindows.get(segment);
+  }
+
+  public RemoteWriteResult<SegmentPartition> preCommit(
+      final int kafkaPartition
+  ) {
+    final var partitionSegments = kafkaPartitionToSegments.get(kafkaPartition);
+
+    final var segmentBatch = partitionSegments.segmentBatch;
+    final SegmentRoll pendingRoll = partitioner.rolledSegments(
+        name, segmentBatch.flushedStreamTime, segmentBatch.batchStreamTime
+    );
+
+    segmentBatch.prepareRoll(pendingRoll);
+
+    for (final long segmentId : pendingRoll.segmentsToCreate) {
+      partitionSegments.createSegment(new SegmentPartition(kafkaPartition, segmentId));
```

Comment:
this is the only side effect in `preCommit`, right? should be safe if there's a race condition so not a concern

Reply:
yep!